### PR TITLE
Correct processing of AES-SHA stitched ciphers

### PR DESCRIPTION
**Pull request by beldmit**
Fixes: #15706

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Review by paulidale — COMMENTED** [review]
A test case that exercises the failure condition should be added.

**Comment by beldmit** [comment]
> A test case that exercises the failure condition should be added.

Not the exercise I like, but will try :)

**Comment by t8m** [comment]
I've already started on that exercise but got distracted by something else so please tell me how far did you get.

**Comment by beldmit** [comment]
@t8m, didn't start yet, had a lunch :)

**Comment by t8m** [comment]
After some fiddling with test_evp to add that testcase, I see it is much harder than expected. Especially given no documentation for these stitched ciphers. @mattcaswell could you give some hints how they should be used? I know I have to set the TLS1_AAD and the MAC_KEY but then I do not know how to get the whole ciphertext.

**Comment by mattcaswell** [comment]
You need to set `OSSL_CIPHER_PARAM_AEAD_TLS1_AAD`, `OSSL_CIPHER_PARAM_AEAD_MAC_KEY` and  `OSSL_CIPHER_PARAM_TLS_VERSION`. Encryption is done "in place". You just call `EVP_EncryptUpdate`, and the output buffer should be the same as the input. Each record is a new `EVP_EncryptUpdate` call. The critical code that does all of this is here:

https://github.com/openssl/openssl/blob/f77208693ec3bda99618e6f76c0f8d279c0077bb/ssl/t1_enc.c#L137-L172
and here:
https://github.com/openssl/openssl/blob/f77208693ec3bda99618e6f76c0f8d279c0077bb/ssl/record/ssl3_record.c#L936-L1272

File: providers/implementations/ciphers/cipher_aes_cbc_hmac_sha256_hw.c [hunk 1]
```diff
@@ -733,7 +733,7 @@ static int aesni_cbc_hmac_sha256_set_tls1_aad(void *vctx,
             if (len < AES_BLOCK_SIZE)
                 return 0;
             len -= AES_BLOCK_SIZE;
-            p[aad_len] = len >> 8;
+            p[aad_len - 2] = len >> 8;
```
Comment (slontis) [inline providers/implementations/ciphers/cipher_aes_cbc_hmac_sha256_hw.c:736]: Good find.. Obviously it should match line 727..

Comment (beldmit) [reply providers/implementations/ciphers/cipher_aes_cbc_hmac_sha256_hw.c:736]: All the other stitched ciphers look good

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
For expediency, this could be merged and a test provided in a followup PR.

**Comment by beldmit** [comment]
Let's discuss it

**Comment by t-j-h** [comment]
OTC decision to proceed with PR without the test and have adding the test in  another PR.

**Comment by paulidale** [comment]
OTC: adding the test later is fine.

**Comment by beldmit** [comment]
Should I do anything FIPS-related action on commit (e.g. `make update` or smth else)?

**Comment by t8m** [comment]
> Should I do anything FIPS-related action on commit (e.g. `make update` or smth else)?

Nope. That is not needed anymore.

**Review by t8m — APPROVED** [review]

**Review by slontis — APPROVED** [review]

**Comment by paulidale** [comment]
Merged.

**Pull request closed without merging**